PP-747: Modify the sched object to have additional options to configu…

suresh-thelkar opened the pull request:
…re pbs_sched

<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-747](https://pbspro.atlassian.net/browse/PP-747)**

#### Problem description
* This ticket is one of the major user-stories of the following EPIC.
**[PP-337](https://pbspro.atlassian.net/browse/PP-337)** Multiple schedulers servicing the PBS cluster

The current user-story actually implements major interface for the MultiSched feature. After this we can actually run multiple instances of scheduler each serving a set of partitions in parallel which actually accomplishes the following goal.

_As clusters get larger and workloads vary it is becoming critical that the jobs get evaluated in as short as time possible to ensure that the correct workload is being run. Using multiple schedulers to address this issue can allow for different scheduling policies and quicker turnaround time for large number of jobs or nodes._

#### Solution description
Nodes and Queues in PBS now have a special attribute called partition. This can be useful in combining or grouping a set of nodes and queues together.
For example we can set partition attribute on nodes and queues as follows.
```
qmgr -c "set node n1 partition=p1"
qmgr -c "set node n2 partition=p1"
qmgr -c "set queue q1 partition = p1"
```
Here we grouped n1,n2 and q1 under partition p1. Similarly we can come up with multiple partitions as per our requirements. Let us say we have one more partition p2 which groups nodes n3,n4,n5 and q2.

A scheduler other than the default scheduler can be associated to one or more partitions as given in the following example.
`qmgr -c "set sched sc1 partition=p1"`
To associate more than one partition,
`qmgr -c "set sched sc1 partition += p2"`

Now the scheduler sc1 can serve two partitions p1 and p2.

An important note here is we should configure the scheduler first in the server then only it can be started. If a server does not know a scheduler id then server rejects it and that scheduler won't come up.

If we try to start a scheduler which is not associated with at least one partition then the following error is thrown while it is coming up.
         "Scheduler should contain at least single partition"
Scheduler will only come up only if it is fully configured at the server.

Once the scheduler comes up successfully it is ready to serve partitions p1, p2 in this case.  All the scheduler features should work as expected. Some of them are as follows

* We can submit jobs to q1 which are handled by sc1 scheduler in this case.
    Example: qsub -q q1 -- /bin/sleep 100
    This job should be served by sc1 and we can observe the same in the log file in sched_logs_sc1 
    directory.  We can also observe the same in pbsnodes -av output i.e. this job should only go to one of 
    the nodes that is part of the partition p1.  
* qrun of a job submitted to the above queue or partition should work.
* suspend/resume functionality should work.
* Array jobs can also be submitted and its functionality should work.
* One can create resources of various types and each scheduler should work as expected and schedule 
   using these resources.
* Reservations until we implement next interface will now only be served by default scheduler.
* Have done all the code changes in such a way that backward compatibility with the default scheduler is 
   maintained.
* server's scheduler_iteration and scheduling attributes are still kept even though we have these 
  attributes for the scheduler object also. This is done to ensure the backward compatibility. If anybody 
  sets these attributes at one object then they are automatically reflected in another object.
* Certain operations like changing the scheduler port and sched_priv directory etc are not allowed on 
  default scheduler to maintain the backward compatibility.
* Dynamically one can associate multiple partitions to the scheduler. We no need to restart or SIGHUP 
   the corresponding scheduler.
* Dynamically one can dissociate a partition from the scheduler.
* One can change sched_priv/sched_log directory dynamically without restarting or SIGHUPing the 
  scheduler.
* If a partition is associated with one scheduler then it can not be associated to any other scheduler.
* Scheduler at any point in time if it fails in accepting the new configuration then it switches back to its 
   old configuration.
* All operations for queues that are having no partitions should be served by default scheduler.
#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

arungrover commented on src/server/hook_func.c:
```diff
@@ -4305,7 +4305,7 @@ int server_process_hooks(int rq_type, char *rq_user, char *rq_host, hook *phook,
 	if (pbs_python_get_scheduler_restart_cycle_flag() == TRUE) {
 
-		set_scheduler_flag(SCH_SCHEDULE_RESTART_CYCLE);
+		set_scheduler_flag(SCH_SCHEDULE_RESTART_CYCLE, dflt_scheduler);
```
We will need to know which scheduler's sched cycle needs to be restarted. I guess this comment does not apply to the current pull but it needs to be addressed in future.

suresh-thelkar replied:
@arungrover 
What you observed is right. This does not apply to the current pull request where as we can address this as part of subsequent interfaces. For the time-being only default scheduler is restarted from the hooks interface.

arungrover commented on src/server/node_manager.c:
```diff
@@ -8693,7 +8701,12 @@ free_sister_vnodes(job *pjob, char *vnodelist, char *err_msg,
+	if (find_assoc_sched_pj(pjob, &psched))
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to reach scheduler associated with job %s", pjob->ji_qs.ji_jobid);
```
Please update it to "Unable to find scheduler associated with partition"

suresh-thelkar replied:
Done

arungrover commented on src/server/node_manager.c:
```diff
@@ -3040,7 +3042,12 @@ deallocate_job(mominfo_t *pmom, job *pjob)
+	if (find_assoc_sched_pj(pjob, &psched))
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to reach scheduler associated with job %s", pjob->ji_qs.ji_jobid);
```
Please update it to "Unable to find scheduler associated with partition"

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_init.c:
```diff
@@ -733,10 +737,11 @@ pbsd_init(int type)
 		}
+		svr_save_db(&server, SVR_SAVE_NEW);
+		dflt_scheduler = sched_alloc(PBS_DFLT_SCHED_NAME);
+		(void)sched_save_db(dflt_scheduler, SVR_SAVE_NEW);
```
I think I gave this comment last time when a pull was raised related to multi scheduler - Shouldn't we set the default and then save it?

suresh-thelkar replied:
@arungrover 
I think we also had discussed about the same earlier. Please find the following link for more details.
https://github.com/PBSPro/pbspro/pull/324#discussion_r125700738

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2045,53 +2047,56 @@ main(int argc, char **argv)
+					schedule_high(psched);
+
+
+				if (psched->svr_do_schedule == SCH_SCHEDULE_RESTART_CYCLE) {
```
do we need to check for scheduler state as down here?

suresh-thelkar replied:
This additional check is not required and I removed it.

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2045,53 +2047,56 @@ main(int argc, char **argv)
+							log_event(PBSEVENT_DEBUG2,
+								PBS_EVENTCLASS_SERVER,
+								LOG_NOTICE, msg_daemonname,
+								"sent scheduler restart scheduling cycle request");
```
Please name the scheduler here

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2045,53 +2047,56 @@ main(int argc, char **argv)
 							PBS_EVENTCLASS_SERVER,
 							LOG_NOTICE, msg_daemonname,
-							"sent scheduler restart scheduling cycle request");
+							"no valid secondary connection to scheduler: restart scheduling cycle request ignored");
```
Please name the scheduler here

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2990,9 +2998,9 @@ try_connect_database(pbs_db_conn_t *conn)
 		db_oper_failed_times++;
 		get_db_errmsg(failcode, &db_err_msg);
-		log_set_dberr(db_err_msg, conn->conn_db_err);
+			log_set_dberr(db_err_msg, conn->conn_db_err);
```
indentation is off here

suresh-thelkar replied:
Done

arungrover commented on src/server/pbsd_main.c:
```diff
@@ -2990,9 +2998,9 @@ try_connect_database(pbs_db_conn_t *conn)
+			log_set_dberr(db_err_msg, conn->conn_db_err);
 		log_event(PBSEVENT_SYSTEM | PBSEVENT_FORCE, PBS_EVENTCLASS_SERVER, LOG_CRIT, msg_daemonname, log_buffer);
-		conn->conn_db_state = PBS_DB_DOWN; /* allow to retry to start db again */
+			conn->conn_db_state = PBS_DB_DOWN; /* allow to retry to start db again */
```
indentation is off here

suresh-thelkar replied:
Done

arungrover commented on src/server/queue_func.c:
```diff
@@ -401,8 +402,14 @@ queuestart_action(attribute *pattr, void *pobject, int actmode)
+				if (find_assoc_sched_pq(pque, &psched))
+					set_scheduler_flag(SCH_SCHEDULE_STARTQ, psched);
+				else {
+					sprintf(log_buffer, "No scheduler associated with the queue %s", pque->qu_qs.qu_name);
```
Since schedulers are now associated to partitions, please mention partition in the message (with partition name)

suresh-thelkar replied:
Done

arungrover commented on src/server/req_jobobit.c:
```diff
@@ -649,7 +650,12 @@ rel_resc(job *pjob)
+	if (find_assoc_sched_pj(pjob, &psched))
+		set_scheduler_flag(SCH_SCHEDULE_TERM, psched);
+	else {
+		sprintf(log_buffer, "Unable to reach scheduler associated with job %s", pjob->ji_qs.ji_jobid);
```
In this log also we should mention partition instead of jobid

suresh-thelkar replied:
Implemented.

arungrover commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+ *
+ * @retval - 1 if success
+ * 	   - 0 if fail
+ */
```
I think you don't need this function. You can just use find_assoc_sched_partition() and pass pj->ji_qs.ji_queue to it

suresh-thelkar replied:
Implemented.

suresh-thelkar commented:
@arungrover 
The idea of making additional code changes(as you observed correctly) is to make sure that this feature starts working right from this interface itself so that we can get more and more feedback at the early stage.  I have tried to maintain the consistency and minimize the race conditions. Some of these additional code changes might move to server once we implement the further interface. 

Also I started working on the review comments.

suresh-thelkar commented on src/include/pbs_error.h:
```diff
@@ -292,7 +292,8 @@ extern "C" {
 #define PBSE_ALPS_SWITCH_ERR 15222	/* ALPS failed to do the suspend/resume */
-
+#define PBSE_OP_NOT_PERMITTED 15223 	/* Operation not permitted on default scheduler */
+#define PBSE_PART_ALREADY_USED 15224 	/* Partition already used */
```
Implemented.

suresh-thelkar commented on src/include/pbs_sched.h:
```diff
@@ -88,6 +97,11 @@ typedef struct pbs_sched {
 extern	pbs_list_head	svr_allscheds;
+extern void set_scheduler_flag(int flag, pbs_sched *psched);
+extern int find_assoc_sched_jid(char *jid, pbs_sched **target_sched);
+extern int find_assoc_sched_pj(job *pj, pbs_sched **target_sched);
```
Implemented.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -258,7 +258,7 @@
         <member_at_free>free_null</member_at_free>
         <member_at_action>action_sched_port</member_at_action>
-        <member_at_flags><both>MGR_ONLY_SET</both></member_at_flags>
+        <member_at_flags><both>READ_ONLY | ATR_DFLAG_SSET</both></member_at_flags>
```
@arungrover / @bhroam 
Let me explain what I have implemented with respect to the scheduler updating its attributes like sched_port, sched_host etc to PBS Server. 
* As per the initial design, as a step 1. admin can select a port for the new scheduler, step 2. if this 
  port is not free then scheduler anyway picks up the free port available. 
* Thought of reducing one step for the administrator in such a way that scheduler(other than the 
  default scheduler) itself picks-up one of the available ports starting from 15050.  Wanted to 
  update the document only after reviewers also like this idea.
* So In order to meet the above requirement the new scheduler itself should find out the free 
   port and also has to update this back to the server which means it has to do the following things.
  * Very first time, scheduler should act as a client and then connect to the server. Here I am 
     calling the function cnt2server_extend() to get the connector fd. 
   * It has to update server about the selected port by calling update_svr_schedobj(). In this 
     function pbs_manager() gets called which at least requires manager privileges to accomplish 
     the goal.  At the same time I also wanted to make sure that manager or operator can not 
     change the port. To meet both these requirements at the same time,  I am passing 
     SC_DAEMON in extend parameter for  cnt2server_extend(). This is useful in identifying that the request is from the scheduler(who is also manager and its privilege is beyond that), so others can not modify it. (For the server side code changes please look in to req_connect() function) 
   * After updating the port to the server, Immediately calling pbs_disconnect(connector) to 
     disconnect it as client.  Then scheduler becomes like a server as usual.
   * Even though scheduler acts as a client only for the first time due to the above code changes we are also making sure that the request is actually from the scheduler through which we are taking care of authentication issues as well. Also after the next interface of Multisched feature PBS Server is the one who actually controls scheduler i.e. starting, stopping etc. So this level of authentication is enough as per my knowledge.
* One more approach is admin himself has to find the free port available on the host where scheduler runs and configure the port through qmgr.  If the port is not available then the scheduler won't come up. He again has to select one more port and configure it through qmgr.

Please suggest if you have a better solution other than the above approaches so that I can follow it.

arungrover replied:
@suresh-thelkar Now when I think about it, I find it problematic if scheduler decides the port and reports that back to server. We will have to handle a lot of error conditions and sync issues between the time scheduler object is created and scheduler starts and report back the port number. I like your second option better where scheduler will only start on an admin specified port, it is more precise that way.

suresh-thelkar replied:
@bhroam 
As @arungrover likes the second approach, If you are also fine with it then I can go ahead and implement it accordingly. This decision helps me in unblocking/addressing some of the other review comments also.

bhroam replied:
I agree with Arun about the second approach.  In your next pull request, you will have the server starting schedulers.  When that happens, the server will be telling the scheduler what port to start on.  You'd have to rip out all of this code in the scheduler (not to mention the PTL tests that go with it).  I'm fine with having the admin jump through some extra hoops until your next pull request.

suresh-thelkar replied:
Implemented.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -114,7 +114,7 @@
 	<member_at_free>free_str</member_at_free>
 	<member_at_action>NULL_FUNC</member_at_action>
-	<member_at_flags><both>ATR_DFLAG_OPRD | ATR_DFLAG_MGRD | ATR_DFLAG_SvWR</both></member_at_flags>
+	<member_at_flags><both>ATR_DFLAG_OPRD | ATR_DFLAG_MGRD | ATR_DFLAG_MGWR | ATR_DFLAG_SvWR</both></member_at_flags>
```
Please look into my detailed reply to this in the below comment.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -386,7 +386,7 @@
 	<member_at_free>free_str</member_at_free>
 	<member_at_action>NULL_FUNC</member_at_action>
-	<member_at_flags><both>READ_ONLY</both></member_at_flags>
+	<member_at_flags><both>NO_USER_SET | ATR_DFLAG_NOSAVM</both></member_at_flags>
```
@arungrover 
Till we implement the next interface where sever itself controls scheduler i.e. starting, stopping etc we need to allow the setting of this attribute by operator and manager. The reason is given below.

Let us say if admin has configured 5 schedulers but he wants to use only two schedulers at the moment. In this case if "state" is read-only then server keeps on contacting each scheduler i.e. keep calling contact_sched() function infinitely until it succeeds which is as per the existing design only. So to avoid this I have made "state" as writable.  I also noticed that this is in line with the "state" attribute of the node. Thought of updating the EDD only after reviewers also like this idea.

bhroam replied:
@suresh-thelkar I thought the point of the scheduling attribute was to control this.  Why would the server keep trying to contact a scheduler if the scheduling attribute was set to false?

bhroam replied:
Oh, and the node state is not like the scheduler state.  There are node states that the admin sets (e.g., offline).  There is no such scheduling state.  PBS is in complete control of the scheduling state.

suresh-thelkar replied:
@bhroam 
Server does not contact the corresponding schedulers if scheduling is false.
As you suggested, I will use the scheduling attribute itself to control this.

suresh-thelkar replied:
Also I have done code changes accordingly.

suresh-thelkar commented on src/lib/Liblog/pbs_messages.c:
```diff
@@ -405,6 +405,8 @@ char *msg_cannot_set_route_que = "Route queues are incompatible with the partiti
 char *msg_queue_not_in_partition = "Queue %s is not part of partition for node";
 char *msg_partition_not_in_queue = "Partition %s is not part of queue for node";
 char *msg_invalid_partion_in_queue = "Invalid partition in queue";
+char *msg_op_not_permitted = "Operation is not permitted on default scheduler";
```
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -118,12 +122,39 @@
 static prev_job_info *last_running = NULL;
 static int last_running_size = 0;
 
+char	scheduler_name[PBS_MAXHOSTNAME+1] = "Me";  /*arbitrary string*/
```
It is actually scheduler_host_name only. I renamed it.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -520,6 +551,7 @@ init_scheduling_cycle(status *policy, int pbs_sd, server_info *sinfo)
 int
 schedule(int cmd, int sd, char *runjobid)
 {
+	update_svr_sched_state(SC_SCHEDULING);
```
@arungrover / @bhroam 
As we all know scheduler can have one of the three states "down", "idle" and "scheduling".  Marking the scheduler state to either "down" or  "idle" from the server is straightforward because server knows when to kick off scheduling cycle.  To mark the state to "scheduling" is not trivial in the scenarios like the following.
* Let us say there are huge number of jobs in the system(like millions of jobs) then we really don't know how much time scheduler takes to complete scheduling them. During this time we should keep the state of scheduler as "scheduling" only.  This is possible only when scheduler updates to server that it has finished scheduling.  In the absence of this as per my knowledge I think server has no way to know that a scheduling cycle is over at so and so time. 
* Since most of the scheduler attributes can now be configured through qmgr during which an admin can make a mistake and scheduler might go down because of this.  In this case code changes are added to update server before scheduler actually goes down. This way there are less chances of race conditions.  But if we want to find this out from the server itself then it might take more time for it when compared to scheduler updating its own state to server.

Due to the above reasons I thought scheduler is the right person to update its state rather than somebody like server finding out the same.

bhroam replied:
@suresh-thelkar The server has all the information it needs to update the scheduling attribute.  It did it before, and it still has it.  The server's state would update to scheduling during a scheduling cycle and turn back to Active when the cycle was over.  The server opens the connection to the scheduler.  At this point you update the state to scheduling.  When the scheduler closes the connection, the server updates the state to idle.

suresh-thelkar replied:
@bhroam 
Thanks for your suggestion. I have made code changes accordingly.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+		int err;
+		int priv_dir_update_fail = 0;
+		if ((log_dir != NULL) && strcmp(log_dir, tmp_log_dir) != 0) {
+			(void)snprintf(path_log,  MAXPATHLEN, tmp_log_dir);
```
MAXPATHLEN is not same is _POSIX_PATH_MAX or _MAX_PATH. I think it is better to redefine MAXPATHLEN as follows in win.h
#define MAXPATHLEN _MAX_PATH
Hope this is fine.

arungrover replied:
It is okay as long as all other places where it is being used do not break. I'd rather limit the scope of the change to just this place and use sizeof operator

suresh-thelkar replied:
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+			log_close(1);
+			if (log_open(logfile, path_log) == -1) {
+				struct attropl *patt;
+				/* update the sched_log attribute with its previous value only */
```
If we don't update the server back then there is a mismatch of qmgr -c "p sched" before and after the configuration update in case if the new configuration is wrong.
Example: Admin wants to change sched_log directory to /tmp/sched_log. Lets us this directory does not exist, then scheduler switches back to using old directory, but the problem is who will update server back that scheduler is still using the old directory.
Before change:
qmgr -c "p sched" |grep sched_logs|grep sc1
set sched sc1 sched_log = /var/spool/pbs/sched_logs_sc1

After change:
qmgr -c "p sched" |grep sched_logs|grep sc1
set sched sc1 sched_log = /tmp/sched_logs_sc1

Due to the above reasons I think listing or printing scheduler should always show the accurate information. 
Updating the comment of scheduler is not needed here since scheduler started using its previous log directory in case of a failure. I need to update the design document after review.

bhroam replied:
I disagree that we should set the anything on the server back.  I agree that the server should always have accurate information.   We don't second guess the admin.  If the admin set something, we don't want to set it to anything.  The admin told us what they want.  If they set something in error, then we stop scheduling and set the scheduling state to down.  If I recall correctly, this is what the EDD actually says.

arungrover replied:
I think in this case as well I will go by what admin is saying and report error from scheduler if the directories are not accessible. I also think that you please update the design document first because this may lead to discussions and who knows, maybe you don't have to change code at all.

suresh-thelkar replied:
Implemented. If Admin sets something wrong then the scheduler comment is updated with the error message and also as @bhroam suggested, making scheduling to false in this case.
Also updated EDD with this change.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+				if (attribs == NULL) {
+					sprintf(log_buffer, "can't update scheduler attribs, calloc failed");
+					log_err(-1, __func__, log_buffer);
+					return;
```
These variables are freed in COPY_ATTR_VALUE macro.

suresh-thelkar replied:
@arungrover 
Please look into the code at around line numbers 2446 in the else block I am calling free. Similarly calling free() for sched_priv case.
else {
				free(log_dir);
				log_dir = tmp_log_dir;
				sprintf(log_buffer, "scheduler log directory is changed to %s", log_dir);
				schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_SCHED, LOG_INFO,
						"reconfigure", log_buffer);
}

arungrover replied:
I think there is some confusion here.
- Line 2446 in the file shows 	`COPY_ATTR_VALUE(tmp_log_dir, attr->value);`
- the comment is about freeing the tmp_* variables not about log_dir or priv_dir

suresh-thelkar replied:
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+						"reconfigure", log_buffer);
+			}
+		} else
+			log_dir = tmp_log_dir;
```
Thanks for catching this. This statement is not required.

suresh-thelkar replied:
@arungrover 
I have recollected why I have set log_dir again. The above else block is needed and gets executed only once to store the log_dir value from tmp_log_dir. If we do not have this change then we can not change the sched_logs directory from qmgr.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ *
+ * @retval Error code
+ * @return -1 - Failure
```
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+	sched_settings_frm_svr(ss);
+
+	if (!dflt_sched && (partitions == NULL)) {
+		sprintf(log_buffer, "Scheduler does not contain a partition. shutting down");
```
If there are no partitions associated to a scheduler then it is exactly identical to default scheduler and hence there is no need of running both schedulers due to which we are throwing the above error message and shutting down the scheduler.  I have updated the design document accordingly.

suresh-thelkar replied:
No, partition is not the one that is used to identify whether it is the default scheduler or the other scheduler.  I thought, running two schedulers which serves the same set of nodes which do not have partitions is not needed and hence the above change.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -220,6 +221,8 @@ die(int sig)
 
 	log_close(1);
+	if (connector > 0)
+		update_svr_sched_state(SC_DOWN);
```
Since now we are finding out the state of the scheduler in server itself, I have removed this code change.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1045,13 +998,28 @@ main(int argc, char *argv[])
+	if (dflt_sched) {
+		(void)sprintf(log_buffer, "%s/sched_priv", pbs_conf.pbs_home_path);
+	} else {
+		(void)sprintf(log_buffer, "%s/sched_priv_%s", pbs_conf.pbs_home_path, sc_name);
```
yes

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+	if (pq == NULL)
+		return 0;
+
+	if (pq->qu_attr[QA_ATR_partition].at_flags & ATR_VFLAG_SET) {
```
Done

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+pbs_sched *
+find_sched_from_sock(int sock)
+{
+	pbs_sched *psched;
```
Done

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -311,24 +505,24 @@ schedule_jobs()
 		while (pdefr) {
 			if (pdefr->dr_sent == 0) {
-				svr_do_schedule = SCH_SCHEDULE_AJOB;
+				dflt_scheduler->svr_do_schedule = SCH_SCHEDULE_AJOB;
```
@bhroam 
qrun in general scenario other than this one is actually working fine. Thanks for catching this, I have modified the code so as to take care of this as well.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -360,13 +554,16 @@ static void
+	set_sched_sock(-1, psched);
 
 	/* clear list of jobs which were altered/modified during cycle */
 	am_jobs.am_used = 0;
```
@bhroam 
I have implemented code changes for your earlier comment due to which I hope nothing needs to be done in this function. Please let me know if anything else is required here.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -449,20 +646,34 @@ was_job_alteredmoved(job *pjob)
-		svr_do_schedule = flag;
+		single_sched = 0;
+
+	for (single_sched || (psched = (pbs_sched*) GET_NEXT(svr_allscheds)); psched ;psched = (pbs_sched*) GET_NEXT(psched->sc_link)) {
```
Done

suresh-thelkar commented on src/include/pbs_error.h:
```diff
@@ -292,7 +292,8 @@ extern "C" {
 #define PBSE_INVALID_PARTITION_QUE 15221 /* Invalid partition to the queue */
 #define PBSE_ALPS_SWITCH_ERR 15222	/* ALPS failed to do the suspend/resume */
-
+#define PBSE_OP_NOT_PERMITTED 15223 	/* Operation not permitted on default scheduler */
```
I have changed this accordingly.

suresh-thelkar commented on src/include/pbs_error.h:
```diff
@@ -292,7 +292,8 @@ extern "C" {
 #define PBSE_ALPS_SWITCH_ERR 15222	/* ALPS failed to do the suspend/resume */
-
+#define PBSE_OP_NOT_PERMITTED 15223 	/* Operation not permitted on default scheduler */
+#define PBSE_PART_ALREADY_USED 15224 	/* Partition already used */
```
I have renamed it to PBSE_SCHED_PARTITION_ALREADY_EXISTS.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -114,7 +114,7 @@
 	<member_at_free>free_str</member_at_free>
 	<member_at_action>NULL_FUNC</member_at_action>
-	<member_at_flags><both>ATR_DFLAG_OPRD | ATR_DFLAG_MGRD | ATR_DFLAG_SvWR</both></member_at_flags>
+	<member_at_flags><both>ATR_DFLAG_OPRD | ATR_DFLAG_MGRD | ATR_DFLAG_MGWR | ATR_DFLAG_SvWR</both></member_at_flags>
```
I can address this comment accordingly after getting the reviewers opinion on the following.
https://github.com/PBSPro/pbspro/pull/454#discussion_r151905571

bhroam replied:
Right now the version is communicated from the scheduler to the server on the first scheduling cycle.  You should be able to continue to do this.

suresh-thelkar replied:
Have modified the code accordingly.

suresh-thelkar replied:
These code changes were required initially when scheduler itself was updating its port back to the server. Since I have modified this now in such a way that admin himself will give sched_port/sched_host through qmgr the above code changes are not required. I just forgot to commit the above file. Please look into my latest commits which I am going to push shortly in a new pull request for the changes.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -270,10 +270,10 @@
 	<member_at_decode>decode_arst</member_at_decode>
 	<member_at_encode>encode_arst</member_at_encode>
-	<member_at_set>set_arst</member_at_set>
+	<member_at_set>set_arst_uniq</member_at_set>
```
We are just making sure that no duplicate value is allowed for this attribute. Actual data validation of this attribute is happening in action_sched_port() function and hence I think it is fine to use set_arst_uniq() here.  I have also observed that similar mechanism is followed for ATTR_NODE_Mom("Mom") attribute of the node. Hope this is fine.

bhroam replied:
You are correct, this is existing behavior for the mom attribute.  I personally disagree with its use, but since the code is already there, go ahead and use it.

suresh-thelkar commented on src/lib/Libattr/master_sched_attr_def.xml:
```diff
@@ -386,7 +386,7 @@
 	<member_at_free>free_str</member_at_free>
 	<member_at_action>NULL_FUNC</member_at_action>
-	<member_at_flags><both>READ_ONLY</both></member_at_flags>
+	<member_at_flags><both>NO_USER_SET | ATR_DFLAG_NOSAVM</both></member_at_flags>
```
Have changed the permissions to READ_ONLY | ATR_DFLAG_SSET.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -118,12 +122,39 @@
+char	sc_name[PBS_MAXSCHEDNAME];
+char	*log_dir = NULL;
+char	*priv_dir = NULL;
+char	*partitions = NULL;
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -118,12 +122,39 @@
+char	path_log[_POSIX_PATH_MAX];
+#endif
+int	dflt_sched = 0;
+
```
Moved the required variables to globals.c

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -118,12 +122,39 @@
+		DEST = (char*)malloc(len + 1); \
+		strncpy(DEST, attr->value, len); \
+		DEST[len] = '\0';\
+	}
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+void
+update_svr_sched_state(char *state)
```
I have changed this code as per your earlier comment where in I have moved these code changes to server itself.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+static void
+sched_settings_frm_svr(struct batch_status *status)
```
This is a static function and we are using it only in fifo.c. Hope this is fine.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+	attr = status->attribs;
+	/*
+	 * resetting the following before fetching from batch_status.
+	 */
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+				attribs = (struct  attropl *)calloc(1, sizeof(struct attropl));
+				if (attribs == NULL) {
+					sprintf(log_buffer, "can't update scheduler attribs, calloc failed");
+					log_err(-1, __func__, log_buffer);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+				patt->next = NULL;
+				err = pbs_manager(connector,
+					MGR_CMD_SET, MGR_OBJ_SCHED,
+					sc_name, attribs, NULL);
```
This functionality might change as per earlier comment and this piece of code may not be required.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ * @retval Error code
+ * @return -1 - Failure
+ * @return  0 - Success
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+	struct batch_status *ss = NULL;
+
+	/* This command is only sent on restart of the server */
+	if (cmd != 0 && cmd == SCH_SCHEDULE_FIRST)
```
Changed to macro.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+	if (cmd != 0 && cmd == SCH_SCHEDULE_FIRST)
+		svr_knows_me = 0;
+
+	if (cmd !=0 && svr_knows_me || cmd == SCH_ERROR || connector < 0)
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -108,13 +107,17 @@
 #include	"pbs_ecl.h"
 #include	"pbs_share.h"
+#include "config.h"
+#include "fifo.h"
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1045,13 +998,28 @@ main(int argc, char *argv[])
 	}
+	/* Check whether the sched name is specified in command line args
+	 * If specified then copy the sched name
+	 * */
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1045,13 +998,28 @@ main(int argc, char *argv[])
+	else {
+		strncpy(sc_name, PBS_DFLT_SCHED_NAME, PBS_MAXSCHEDNAME -1);
+		dflt_sched = 1;
+	}
```
At present we can also run the scheduler in either of the following ways
Example:
    pbs_sched sc1 -d /var/spool/pbs    
    pbs_sched -d /var/spool/pbs sc1

where sc1 is name of the scheduler. If new option required still, I can go ahead and do this.

bhroam replied:
No other pbs daemon takes an argument like this.  Setting the scheduler name should be a getopt() option.

suresh-thelkar replied:
I have added new option to give the name of scheduler which "-I". one can start the scheduler as follows going forward
pbs_sched -I sc1 -S 15050

bhroam replied:
This is fine, but I'm just curious.  Why -l instead of -n?

suresh-thelkar replied:
Since capital letter 'S' is used for port, I initially thought of using capital letter 'N' but this is already reserved and hence used 'I' (stands for scheduler ID).  Please let me know if I still need to use '-n'.

bhroam replied:
My only worry is that we call it a name everywhere else.  If we call it an id here, it could be confusing.  I'll leave it up to you to decide what is best.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -554,17 +586,31 @@ schedule(int cmd, int sd, char *runjobid)
 			reset_global_resource_ptrs();
```
@Thanks Arun for your suggestion. Have created a new scheduler command SCH_ATTRS_CONFIGURE and using this in case of resource updates.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+					sc_name, attribs, NULL);
+				free(attribs);
+				if (err) {
+					sprintf(log_buffer, "Failed to update log_dir value %s at the server", log_dir);
```
This code now is not necessary  and hence removed as we are going to set the scheduler comment in case of any failure to access the log directory.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+					return;
+				}
+				sprintf(log_buffer, "%s: logfile could not be opened under directory %s", logfile, tmp_log_dir);
+				sprintf(log_buffer, "switching back to previous directory %s", log_dir);
```
This code now is not necessary and hence removed as we are going to set the scheduler comment in case of any failure to access the log directory.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+				log_dir = tmp_log_dir;
+				sprintf(log_buffer, "scheduler log directory is changed to %s", log_dir);
+				schdlog(PBSEVENT_SCHED, PBS_EVENTCLASS_SCHED, LOG_INFO,
+						"reconfigure", log_buffer);
```
This code now is not necessary and hence removed as we are going to set the scheduler comment in case of any failure to access the log directory.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+				c |= chk_file_sec(pbs_conf.pbs_environment, 0, 0, S_IWGRP|S_IWOTH, 0);
+				if (c != 0) {
+					sprintf(log_buffer, "PBS failed validation checks for directory %s", tmp_priv_dir);
+					sprintf(log_buffer, "switching back to previous directory %s", priv_dir);
```
Same as the above response.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,306 @@ next_job(status *policy, server_info *sinfo, int flag)
+				}
+			#endif  /* not DEBUG and not NO_SECURITY_CHECK */
+			if (c == 0) {
+				if (chdir(log_buffer) == -1) {
```
ame as the above response.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+				if (chdir(log_buffer) == -1) {
+					sprintf(log_buffer, "PBS failed validation checks for directory %s", tmp_priv_dir);
+					sprintf(log_buffer, "switching back to previous directory %s", priv_dir);
+					log_err(-1, __func__, log_buffer);
```
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2322,3 +2369,307 @@ next_job(status *policy, server_info *sinfo, int flag)
+			err = pbs_manager(connector,
+				MGR_CMD_SET, MGR_OBJ_SCHED,
+				sc_name, attribs, NULL);
+			free(attribs);
```
Now scheduler's comment is updated with the reason for failure.  Scheduler's other attributes are not changed. Admin can still see what is the value he is trying to set last time.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1171,12 +1143,33 @@ main(int argc, char *argv[])
+		}
+	} else {
+		int try_port;
+		for (try_port=STARTING_PORT_NUM;try_port < MAX_PORT_NUM;try_port++) {
```
The above code is not needed now.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1336,6 +1329,15 @@ main(int argc, char *argv[])
 	sprintf(log_buffer, "%s startup pid %ld", argv[0], (long)pid);
 	log_record(PBSEVENT_SYSTEM, PBS_EVENTCLASS_SERVER, LOG_INFO, __func__, log_buffer);
 
+	/*connector = pbs_connect_noblk(NULL,10);*/
```
The above code is not needed now.

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1417,12 +1419,13 @@ main(int argc, char *argv[])
 		if (!FD_ISSET(server_sock, &fdset))
 			continue;
 
+		pbs_disconnect(connector);
```
The above code is not needed now.

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1138,6 +1072,16 @@ main(int argc, char *argv[])
+	if (argc > optind)
+		strncpy(sc_name, argv[optind], PBS_MAXSCHEDNAME - 1);
+	else {
+		strncpy(sc_name, PBS_DFLT_SCHED_NAME, PBS_MAXSCHEDNAME -1);
```
I have modified this piece of code in such a way that sc_name as a pointer rather than an array and made changes accordingly.

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1318,16 +1269,38 @@ main(int argc, char *argv[])
+		}
+	} else {
+		int try_port;
+		for (try_port=STARTING_PORT_NUM;try_port < MAX_PORT_NUM;try_port++) {
```
The above code is not needed now.

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -1588,11 +1569,15 @@ main(int argc, char *argv[])
 		if (!FD_ISSET(server_sock, &fdset))
 			continue;
 
+		pbs_disconnect(connector);
```
The above code is not needed now.

suresh-thelkar commented on src/server/queue_func.c:
```diff
@@ -401,8 +402,14 @@ queuestart_action(attribute *pattr, void *pobject, int actmode)
+					set_scheduler_flag(SCH_SCHEDULE_STARTQ, psched);
+				else {
+					sprintf(log_buffer, "No scheduler associated with the queue %s", pque->qu_qs.qu_name);
+					log_err(-1, __func__, log_buffer);
```
This message comes in a corner case. Let's say eligible_time_enable attribute is set to true and admin after some time after initial configuration assume that he deleted the scheduler object that  handles this particular queue, then if somebody submits a job to this queue in that scenario the above message comes. Please note that the above message is rephrased as well in the latest commit.

suresh-thelkar commented:
PTL test cases for this PR were initially in the following PR raised by @visheshh.
https://github.com/PBSPro/pbspro/pull/457
 As requested by @arungrover, I have now merged the above PTL test cases also to this PR in a separate commit. 

@bhroam / @arungrover  / @varunsonkar 
Can you please review these PTL test cases as well and provide your comments.

suresh-thelkar commented:
@bhroam / @arungrover 
I have implemented most of the round-1 review comments. Please look into the latest commits for the same. 
Also added code changes to fix SLESS11 crash that we were seeing.  However please note that some of the code changes done to fix the issue are not actually related to Multisched and hence I have raised couple of tickets on master branch to fix them.  But for the interest of time and to test Multisched completely these code changes are done as part of this branch itself.

suresh-thelkar commented on src/scheduler/node_info.c:
```diff
@@ -176,9 +176,10 @@ query_nodes(int pbs_sd, server_info *sinfo)
 
 	/* get nodes from PBS server */
-	if ((nodes = pbs_statvnode(pbs_sd, NULL, NULL, NULL)) == NULL) {
+	if ((nodes = pbs_statvnode(pbs_sd, NULL, NULL, partitions)) == NULL) {
```
According to our last meeting discussion, coming up with new IFL like pbs_selvnodestat() is taken up in the future based on the need(Lot of discussion is needed and can be continued separately on the same).  In the interest of time, As far as possible we are planning to use the existing API itself which actually avoids bumping up of IFL library version.

bhroam replied:
I understand that writing the new selstat calls have been pushed out.  I don't want to use the extend field though.  If you extend the IFL layer (even in a way you don't have to bump the version), that has to be supported.  I'd rather see you query everything and then filter it on the scheduler side.  This means no change to IFL has to be made.

suresh-thelkar commented on src/scheduler/queue_info.c:
```diff
@@ -148,7 +148,7 @@ query_queues(status *policy, int pbs_sd, server_info *sinfo)
 
 	/* get queue info from PBS server */
-	if ((queues = pbs_statque(pbs_sd, NULL, NULL, NULL)) == NULL) {
+	if ((queues = pbs_statque(pbs_sd, NULL, NULL, partitions)) == NULL) {
```
It is similar to the one that I have responded in https://github.com/PBSPro/pbspro/pull/454#discussion_r156557726

suresh-thelkar commented on src/server/req_getcred.c:
```diff
@@ -85,6 +86,9 @@ req_connect(struct batch_request *preq)
 		conn->cn_authen |= PBS_NET_CONN_FROM_QSUB_DAEMON;
+	} else if ( (preq->rq_extend != NULL) && \
+		(strcmp(preq->rq_extend, SC_DAEMON) == 0) ) {
+		conn->cn_authen |= PBS_NET_CONN_FROM_PRIVIL;
```
Done

suresh-thelkar commented on src/server/req_modify.c:
```diff
@@ -242,7 +241,7 @@ req_modifyjob(struct batch_request *preq)
 		 * no need to delay the job until next cycle.
 		 */
-		if ((preq->rq_conn != scheduler_sock) && (scheduler_jobs_stat) && (job_attr_def[i].at_flags & ATR_DFLAG_SCGALT))
+		if ((find_sched_from_sock(preq->rq_conn) == NULL) && (scheduler_jobs_stat) && (job_attr_def[i].at_flags & ATR_DFLAG_SCGALT))
```
Implemented.

suresh-thelkar commented on src/server/req_runjob.c:
```diff
@@ -458,7 +464,8 @@ req_runjob(struct batch_request *preq)
 		net_add_close_func(preq->rq_conn, clear_from_defr);
 
-		if (schedule_jobs() == -1) {
+		if ((strcmp(psched->sch_attr[SCHED_ATR_sched_state].at_val.at_str, SC_DOWN))
```
I have removed the above as per the other comments.

suresh-thelkar commented on src/server/req_select.c:
```diff
@@ -366,7 +365,8 @@ req_selectjobs(struct batch_request *preq)
 	 * instead of a per-queue selstat()
 	 */
-	if ((scheduler_sock != -1) && (preq->rq_conn == scheduler_sock) && (!scheduler_jobs_stat)) {
+	if ((find_sched_from_sock(preq->rq_conn) != NULL) && (!scheduler_jobs_stat)) {
```
@bhroam Can you please elaborate if I need to do any changes here.

bhroam replied:
The way this works now is that if in non-throughput mode a job is modified after the the scheduler stats the jobs, the job can not run in that cycle.  If we are going to modify this functionality for all schedulers, you'll need a scheduler_jobs_stat for each scheduler to keep track of that.  Also there is an array that keeps track of the jobs that were modified.  You'll need one of those per scheduler as well.  While this is not default functionality, it is stable functionality.  Backwards compatibility must be kept.  You could deprecate this functionality, but it you'd need to make sure it still worked with the default scheduler.  Either that or make it work with all schedulers.

suresh-thelkar commented on src/server/req_select.c:
```diff
@@ -366,7 +365,8 @@ req_selectjobs(struct batch_request *preq)
 	 */
-	if ((scheduler_sock != -1) && (preq->rq_conn == scheduler_sock) && (!scheduler_jobs_stat)) {
+	if ((find_sched_from_sock(preq->rq_conn) != NULL) && (!scheduler_jobs_stat)) {
+		//if ((dflt_scheduler->scheduler_sock != -1) && (preq->rq_conn == dflt_scheduler->scheduler_sock) && (!scheduler_jobs_stat)) {
```
Done

suresh-thelkar commented on src/server/req_shutdown.c:
```diff
@@ -266,7 +266,7 @@ req_shutdown(struct batch_request *preq)
 
 	if (shutdown_who & SHUT_WHO_SCHED)
-		(void)contact_sched(SCH_QUIT, NULL);	/* tell scheduler to quit */
+		(void)contact_sched(SCH_QUIT, NULL, dflt_scheduler->pbs_scheduler_addr, dflt_scheduler->pbs_scheduler_port);	/* tell scheduler to quit */
```
Changing the behavior of qterm -s is actually part of the next interface where we need to make sure that all schedulers go down with this command as you suggested. Please let me know if I need to do this here itself. For the time being doing it only for default scheduler.

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -118,8 +118,8 @@ static int bad;
 
-static int  status_que(pbs_queue *, struct batch_request *, pbs_list_head *);
-static int status_node(struct pbsnode *, struct batch_request *, pbs_list_head *);
+static int  status_que(pbs_queue *, struct batch_request *, pbs_list_head *, attribute *attr);
```
Thanks for catching this. I have removed the extra space.

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -442,23 +445,86 @@ req_stat_que(struct batch_request *preq)
+					if ((attr.at_val.at_arst->as_string[k] != NULL)
+							&& (!strcmp(
+									pque->qu_attr[QA_ATR_partition].at_val.at_str,
+									attr.at_val.at_arst->as_string[k]))) {
```
Done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -442,23 +445,86 @@ req_stat_que(struct batch_request *preq)
-					rc = 0;
-				else
-					break;
+			if (preq->rq_extend != NULL) {
```
Done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -442,23 +445,86 @@ req_stat_que(struct batch_request *preq)
-					break;
+			if (preq->rq_extend != NULL) {
+				if (pque->qu_attr[QA_ATR_partition].at_flags & ATR_VFLAG_SET) {
+					/*search for the partition in the attr*/
```
Done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -442,23 +445,86 @@ req_stat_que(struct batch_request *preq)
+						if ((attr.at_val.at_arst->as_string[k] != NULL)
+								&& (!strcmp(
+										pque->qu_attr[QA_ATR_partition].at_val.at_str,
+										attr.at_val.at_arst->as_string[k]))) {
```
Done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -481,10 +547,11 @@ req_stat_que(struct batch_request *preq)
 
 static int
-status_que(pbs_queue *pque, struct batch_request *preq, pbs_list_head *pstathd)
+status_que(pbs_queue *pque, struct batch_request *preq, pbs_list_head *pstathd, attribute *attr)
```
done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -481,10 +547,11 @@ req_stat_que(struct batch_request *preq)
 {
 	struct brp_status *pstat;
 	svrattrl	  *pal;
+	int k;
```
These variables are not required and hence deleted them.

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -578,18 +648,83 @@ req_stat_node(struct batch_request *preq)
+				consider_n = 1;
+			}
+		}
+		if (consider_n || preq->rq_conn != dflt_scheduler->scheduler_sock) {
```
The above "if" condition is used for both of the following cases.
1. When scheduler is querying the nodes 
    In case of non default scheduler if its partition does not match with the node's partition then it 
    won't hit the above if block. It exits the loop earlier i.e. it does not reach to this statement.
2. When user/admin list/print nodes 
    In case of non default scheduler we should actually list all the nodes even if they are associated 
    with partition or not if user/admin does things like qmgr -c "print node @default"

In order to satisfy above use cases I have done the changes accordingly.

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -578,18 +648,83 @@ req_stat_node(struct batch_request *preq)
+					for (k = 0; k < attr.at_val.at_arst->as_usedptr; k++) {
+						if ((attr.at_val.at_arst->as_string[k] != NULL)
+								&& (!strcmp(
+										pnode->nd_attr[ND_ATR_partition].at_val.at_str,
```
Done

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -578,18 +648,83 @@ req_stat_node(struct batch_request *preq)
+					consider_n = 1;
+				}
+			}
+			if (consider_n || preq->rq_conn != dflt_scheduler->scheduler_sock) {
```
The above "if" condition is used for both of the following cases.

1. When scheduler is querying the queues
    In case of non default scheduler if its partition does not match with the queue's partition then 
    it won't hit the above if block. It exits the loop earlier i.e. it does not reach to this statement.
2. When user/admin list/print queues
     In case of non default scheduler we should actually list all the queues even if they are 
     associated with partition or not if user/admin does things like qmgr -c "print queue @default"

In order to satisfy above use cases I have done the changes accordingly.

bhroam replied:
I thought we decided all the filtering was going to happen at the scheduler side.  Won't the first bullet not matter?

suresh-thelkar commented on src/server/req_stat.c:
```diff
@@ -623,12 +758,14 @@ req_stat_node(struct batch_request *preq)
 	svrattrl	  *pal;
 	unsigned long		   old_nd_state = VNODE_UNAVAILABLE;
+	int k;
+	int found;
```
Removed the unneeded variables.

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+ * @brief
+ * 		find_assoc_sched - find the corresponding scheduler which is responsible
+ * 		for handling this job.
+ *
```
Done

suresh-thelkar commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+	if (pq == NULL)
+		return 0;
+
+	if (pq->qu_attr[QA_ATR_partition].at_flags & ATR_VFLAG_SET) {
```
Implemented.

suresh-thelkar commented on src/server/sched_func.c:
```diff
@@ -374,5 +452,89 @@ set_sched_default(pbs_sched* psched)
+	if ((psched->sch_attr[(int) SCHED_ATR_sched_log].at_flags & ATR_VFLAG_SET) == 0) {
+		psched->sch_attr[(int) SCHED_ATR_sched_log].at_val.at_str = malloc(MAXPATHLEN + 1);
+		if (psched->sch_attr[(int) SCHED_ATR_sched_log].at_val.at_str == NULL)
+			return;
```
Added error message in the logs with errno in all the cases wherever required.

suresh-thelkar commented on src/server/sched_func.c:
```diff
@@ -374,5 +452,89 @@ set_sched_default(pbs_sched* psched)
+		if (pattr->at_val.at_arst->as_string[i] == NULL)
+			continue;
+		psched = (pbs_sched*) GET_NEXT(svr_allscheds);
+		while (psched) {
```
Implemented.

suresh-thelkar commented on src/server/svr_jobfunc.c:
```diff
@@ -218,6 +218,23 @@ clear_default_resc(job *pjob)
+ */
+void
+tickle_for_reply(void)
+{
```
I was getting weird compilation issue in Windows. Just to remove this error I have mode the above code here.

suresh-thelkar commented:
@bhroam / @arungrover 
Have implemented remaining comments from round-1 and pushed the commits. Please also look into these changes.

suresh-thelkar commented:
@bhroam / @arungrover 
Requesting you to provide your review comments on the latest commits that we have added recently which includes PTL tests as well so that we could continue to implement the comments quickly and proceed further.

Thanks,
Suresh

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -104,6 +129,7 @@ def test_permission_on_scheduler(self):
                             SCHED,
                             id="testDeleteSched",
                             runas=ROOT_USER)
+        print self.server.schedulers
```
Don't use print.  If you want to log something, use self.logger()

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+             'scheduler_iteration': 600,
+             'state': 'idle',
+             'sched_cycle_length': '00:20:00'}
+        self.server.expect(SCHED, a, id='default', max_attempts=10)
```
When you pass more than one attribute to expect() it does an or.  If you want an and, do op=PTL_AND

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+            self.server.manager(MGR_CMD_SET, SCHED,
+                                {'sched_priv': '/var/spool/pbs/sched_priv'},
+                                runas=ROOT_USER, id='TestCommonSched')
+        except PbsManagerError, e:
```
Use the post 2.6 style of except PbsManagerError as e:

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+            self.server.manager(MGR_CMD_SET, SCHED,
+                                {'sched_log': '/var/spool/pbs/sched_logs'},
+                                runas=ROOT_USER, id='TestCommonSched')
+        except PbsManagerError, e:
```
same as above, use 'as' and for all the other except calls in the test suite

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+        self.server.manager(MGR_CMD_SET, SERVER,
+                            {'scheduling': 'True'},
+                            runas=ROOT_USER)
+        self.server.manager(MGR_CMD_LIST, SCHED, id='default')
```
Again, why do you have to do a list?

bhroam commented on test/tests/interfaces/pbs_sched_interface_test.py:
```diff
@@ -156,9 +182,129 @@ def test_set_and_unset_sched_attributes(self):
+
+    def tearDown(self):
+        self.du.run_cmd(self.server.hostname, [
+                        'pkill', '-9', 'pbs_sched'], sudo=True)
```
Use PTL to bring down the schedulers.  Don't use pkill.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+from tests.functional import *
+
+
+@tags('multisched')
```
I have used a tag which was already used in pbs_sched_interface_test.py , I have not used a new tag . Should i change both to a generic scheduler tag ?

bhroam replied:
I'd talk with the PTL steering committee.  They can provide guidance.  I remember when the new tags were being rolled out that we were not supposed to create new tags ourselves.

visheshh commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -3743,7 +3743,10 @@ def _get_pid(self, inst):
         lock = self._instance_to_lock(inst)
-        path = os.path.join(self.pbs_conf['PBS_HOME'], priv, lock)
+        if isinstance(inst, Scheduler) and 'sched_priv' in inst.attributes:
+            path = os.path.join(inst.attributes.get('sched_priv'), lock)
```
Done

visheshh commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -6623,6 +6631,21 @@ def manager(self, cmd, obj_type, attrib=None, id=None, extend=None,
                     del self.queues[i]
                 if obj_type == MGR_OBJ_RSC and i in self.resources:
                     del self.resources[i]
+                if obj_type == SCHED and i in self.schedulers.keys():
```
Done

bhroam commented on src/server/req_stat.c:
```diff
@@ -478,28 +476,23 @@ req_stat_que(struct batch_request *preq)
 	} else {	/* get status of queues */
-		pque = (pbs_queue *)GET_NEXT(svr_queues);
-		while (pque) {
+		for (pque = (pbs_queue *)GET_NEXT(svr_queues); pque;pque = (pbs_queue *)GET_NEXT(pque->qu_link)) {
```
Add a space after the ';'

suresh-thelkar replied:
This is outdated in the sense that we are now doing the filtering of queues at the scheduler side.

bhroam commented on src/server/run_sched.c:
```diff
@@ -155,6 +158,190 @@ put_sched_cmd(int sock, int cmd, char *jobid)
+ *
+ * @retval - 1 if success
+ * 	   - 0 if fail
+ */
```
Any answer here?  Why did you choose to return by reference when find_queuebyname() returns directly?

bhroam commented on src/server/sched_func.c:
```diff
@@ -439,9 +439,11 @@ set_sched_default(pbs_sched* psched)
 		psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str = malloc(SC_STATUS_LEN + 1);
-		if (psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str == NULL)
+		if (psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str == NULL) {
+			snprintf(log_buffer, LOG_BUF_SIZE-1, "malloc failure (errno %d)", errno);
```
You don't need the -1, snprintf() will handle that for you.  Same goes for the other snprintf() calls.

suresh-thelkar replied:
Done

bhroam commented on src/server/sched_func.c:
```diff
@@ -439,9 +439,11 @@ set_sched_default(pbs_sched* psched)
-		if (psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str == NULL)
+		if (psched->sch_attr[(int) SCHED_ATR_sched_state].at_val.at_str == NULL) {
+			snprintf(log_buffer, LOG_BUF_SIZE-1, "malloc failure (errno %d)", errno);
+			log_err(PBSE_SYSTEM, __func__, log_buffer);
```
The first argument to log_err() is the errno, not a PBSE error number.  If you want to log at PBSE_SYSTEM, you'll need to use a log_event().  You don't need to embed the errno as part of the message, use it as the first argument here.  Same goes for the rest of the messages like this one.

suresh-thelkar replied:
Done

bhroam commented on src/server/sched_func.c:
```diff
@@ -374,5 +452,89 @@ set_sched_default(pbs_sched* psched)
+	int k;
+	if (pobj == dflt_scheduler)
+		return PBSE_OP_NOT_PERMITTED;
+	pin_sched = (pbs_sched*) pobj;
```
Still, add a space before the *

bhroam commented on src/server/sched_func.c:
```diff
@@ -374,5 +452,89 @@ set_sched_default(pbs_sched* psched)
+		return PBSE_OP_NOT_PERMITTED;
+	pin_sched = (pbs_sched*) pobj;
+
+	for (i=0; i < pattr->at_val.at_arst->as_usedptr; ++i) {
```
Still, add spaces around the =

bhroam commented on src/server/svr_chk_owner.c:
```diff
@@ -466,7 +466,8 @@ chk_job_request(char *jobid, struct batch_request *preq, int *rc)
 		if (p2) 
 			*p2 = '\0';
-		strcat(jobid, p1);
+		/*strcat(jobid, p1);*/
```
Remove this commented code.

suresh-thelkar replied:
Done

bhroam commented:
I just had this pull request time out on me twice.  We're getting to the point where we should close it and open a new one so it doesn't become completely unresponsive.

suresh-thelkar commented on src/scheduler/constant.h:
```diff
@@ -150,6 +150,18 @@ enum update_accruetype_mode
+#define COPY_ATTR_VALUE(DEST, SRC) \
+	{ \
+		int len = 0;\
+		if (DEST) { \
```
Done

suresh-thelkar commented on src/scheduler/constant.h:
```diff
@@ -150,6 +150,18 @@ enum update_accruetype_mode
+		DEST = (char*)malloc(len + 1); \
+		strncpy(DEST, attr->value, len); \
+		DEST[len] = '\0';\
+	}
```
Implemented

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -631,6 +602,16 @@ intermediate_schedule(int sd, char *jobid)
+		sprintf(log_buffer, "update_svr_schedobj failed");
+		log_err(-1, __func__, log_buffer);
+		return 0;
+	}
```
Modified the code so as not to consider the return value of update_svr_schedobj.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+    def test_multiple_queue_same_partition(self):
+        """
+        Test multiple queue assosciated with same partition
+        is serviced by same scheduler
```
Both queue's were not assigned to different partition , I changed wq4 to have same partition , Now test is matching the description.

visheshh replied:
and updated the above test case to even test that partition p1 job can run on partition p4 node , if both partitions are serviced by same sched

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2528,13 +2480,9 @@ sched_settings_frm_svr(struct batch_status *status)
 					lockfds = open("sched.lock", O_CREAT|O_WRONLY, 0644);
 					if (lockfds < 0) {
 						sprintf(log_buffer, "PBS failed validation checks for directory %s", tmp_priv_dir);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
-						free(priv_dir);
 						priv_dir = tmp_priv_dir;
-						sprintf(log_buffer, "scheduler priv directory has changed to %s", priv_dir);
+						sprintf(log_buffer, "scheduler priv directory has changed to %s", tmp_priv_dir);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
 			free(attribs);
 			if (err) {
-				sprintf(log_buffer, "Failed in updating priv_dir value %s to the server", priv_dir);
+				sprintf(log_buffer, "Failed to update scheduler comment %s at the server", log_dir);
```
Implemented

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+    def test_job_queue_partition(self):
+        """
+        Test job submitted to a queue assosciated to a partition will land
+        into a node assosciated with that partition.
```
done

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.schedulers['sc1'].log_match(log_msg, max_attempts=10,
+                                                starttime=self.server.ctime)
+        self.server.manager(MGR_CMD_UNSET, SCHED, 'partition',
+                            id="sc2")
```
After testing removing of partitions one by one of sc1 and testing log message , I am testing removal of partition of scheduler sc2 by unsetting scheduler attribute . I am not doing log_match because unsetting sched attr will not send any signal back to scheduler and scheduler log wont show that scheduler does not contain partition.

suresh-thelkar replied:
@visheshh 
I have not looked into the complete context which you were discussing but just wanted to update that unsetting a few sched attributes like sched_log and sched_priv will trigger a command SCH_ATTRS_CONFIGURE to the scheduler.

visheshh replied:
@suresh-thelkar If any unsetting partition from a scheduler can also send sch_attrs_configure to the scheduler , Scheduler does not contain a partition might appear in sched_logs too.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
+			return 0;
 		}
-		if ((partitions != NULL) && strcmp(partitions, tmp_partitions) != 0) {
+		if ((partitions != NULL) && (tmp_partitions != NULL) && strcmp(partitions, tmp_partitions) != 0) {
```
Thanks for letting me know this function and I have modified it accordingly.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
+
+		attribs = (struct  attropl *)calloc(1, sizeof(struct attropl));
+		if (attribs == NULL) {
+			sprintf(log_buffer, "can't update scheduler attribs, calloc failed");
```
Implemented.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
+		free(attribs);
+		free(tmp_comment);
+		if (err) {
+			sprintf(log_buffer, "Failed to update scheduler comment %s at the server", log_dir);
```
Done

suresh-thelkar commented on src/scheduler/fifo.h:
```diff
@@ -43,12 +43,10 @@ extern "C" {
+extern char scheduler_host_name[PBS_MAXHOSTNAME+1];
 extern char *logfile;
 extern char path_log[_POSIX_PATH_MAX];
 extern int dflt_sched;
```
Done

suresh-thelkar commented on src/scheduler/globals.c:
```diff
@@ -156,3 +157,15 @@ resdef **boolres = NULL;
 char *cmp_aoename = NULL;
+
+char *partitions = NULL;
+char scheduler_host_name[PBS_MAXHOSTNAME+1] = "Me";  /*arbitrary string*/
```
Done

suresh-thelkar commented on src/scheduler/globals.c:
```diff
@@ -156,3 +157,15 @@ resdef **boolres = NULL;
+char scheduler_host_name[PBS_MAXHOSTNAME+1] = "Me";  /*arbitrary string*/
+char *sc_name = NULL;
+int sched_port = -1;
+char *logfile = (char *)0;
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -792,8 +790,8 @@ are_we_primary()
-	strncpy(scheduler_name, server_host, sizeof(scheduler_name));
-	scheduler_name [ sizeof(scheduler_name) -1 ] = '\0';
+	strncpy(scheduler_host_name, server_host, sizeof(scheduler_host_name));
+	scheduler_host_name [ sizeof(scheduler_host_name) -1 ] = '\0';
```
Done

suresh-thelkar commented on src/scheduler/pbs_sched_win.c:
```diff
@@ -775,8 +773,8 @@ are_we_primary()
-	strncpy(scheduler_name, server_host, sizeof(scheduler_name));
-	scheduler_name [ sizeof(scheduler_name) -1 ] = '\0';
+	strncpy(scheduler_host_name, server_host, sizeof(scheduler_host_name));
+	scheduler_host_name [ sizeof(scheduler_host_name) -1 ] = '\0';
```
Done

suresh-thelkar commented on src/server/req_manager.c:
```diff
@@ -1711,15 +1711,23 @@ void
-	plist = (svrattrl *)GET_NEXT(preq->rq_ind.rq_manager.rq_attr);
 
+
+	for (tmp_plist = (svrattrl *)GET_NEXT(preq->rq_ind.rq_manager.rq_attr);tmp_plist;tmp_plist = (struct svrattrl *)GET_NEXT(tmp_plist->al_link)) {
```
Done

suresh-thelkar commented on src/server/req_manager.c:
```diff
@@ -1711,15 +1711,23 @@ void
+	for (tmp_plist = (svrattrl *)GET_NEXT(preq->rq_ind.rq_manager.rq_attr);tmp_plist;tmp_plist = (struct svrattrl *)GET_NEXT(tmp_plist->al_link)) {
+		if (strcasecmp(tmp_plist->al_name, ATTR_sched_log) == 0 ||
+			strcasecmp(tmp_plist->al_name, ATTR_sched_priv) == 0) {
+			set_scheduler_flag(SCH_ATTRS_CONFIGURE, psched);
```
I would like to leave this as is at the moment. As per my knowledge in future also we are moving towards horizontal scaling of servers but not vertical scaling. Hope this is fine for now.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2460,49 +2407,54 @@ sched_settings_frm_svr(struct batch_status *status)
-				/* update the sched_log attribute with its previous value only */
-				attribs = (struct  attropl *)calloc(1, sizeof(struct attropl));
+				/* update the sched comment attribute with the reason for failure */
+				attribs = (struct  attropl *)calloc(2, sizeof(struct attropl));
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2460,49 +2407,54 @@ sched_settings_frm_svr(struct batch_status *status)
+		char comment[MAX_LOG_SIZE] = {0};
 		if ((log_dir != NULL) && strcmp(log_dir, tmp_log_dir) != 0) {
-			(void)snprintf(path_log,  MAXPATHLEN, tmp_log_dir);
+			(void)snprintf(path_log,  sizeof(path_log), tmp_log_dir);
```
Done

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2452,6 +2397,8 @@ sched_settings_frm_svr(struct batch_status *status)
 			} else if (!strcmp(attr->name, ATTR_partition)) {
 				COPY_ATTR_VALUE(tmp_partitions, attr->value);
+			} else if (!strcmp(attr->name, ATTR_comment)) {
+				COPY_ATTR_VALUE(tmp_comment, attr->value);
```
This part of code is required to clear the comment of a corresponding scheduler at the server in the scenarios like the following.
1. Admin tries to set sched_log to a directory which does not exist for example. Then the scheduler comment is updated with the comment given below.
Qmgr: s sched sc1 sched_log = /tmp/a
Qmgr: l sc sc1
Sched sc1
    sched_host = stblr3
    pbs_version = 17.1.0
    sched_cycle_length = 00:20:00
    sched_port = 15050
    partition = p1
    sched_priv = /var/spool/pbs/suresh_priv_sc1
    sched_log = /tmp/a
    scheduling = False
    scheduler_iteration = 600
    comment = Unable to change the sched_logs directory
    state = idle
2. When admin tries to set the correct value for sched_log again then we need to clear the above comment.  
We need to clear the comment only if it is set earlier otherwise every time the user configures scheduler object correctly we need to clear the comment.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2544,44 +2492,86 @@ sched_settings_frm_svr(struct batch_status *status)
-			/* update the sched_log attribute with its previous value only */
-			attribs = (struct  attropl *)calloc(1, sizeof(struct attropl));
+			/* update the sched comment attribute with the reason for failure */
+			attribs = (struct  attropl *)calloc(2, sizeof(struct attropl));
```
I am freeing the attribs right after the calls to pbs_manager(). Please look into the latest commits.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc5")
+        self.server.manager(MGR_CMD_LIST, SCHED, id="sc5")
+        self.assertEqual(sched.attributes['state'], 'scheduling')
```
Changed to expect and running 500 jobs , 100 is too fast .

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+            self.server.submit(j)
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc5")
+        self.server.manager(MGR_CMD_LIST, SCHED, id="sc5")
```
Not if i use expect , I am removing it

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, SCHED, a, id="sc5")
+        sched = self.server.schedulers['sc5']
+        # Try starting without sched_priv and sched_logs
+        ret = self.server.schedulers['sc5'].start(sched_port='15055')
```
scheduler was always a part of server , It was made as dictionary in PP-851 commit.

bhroam replied:
This is horribly confusing.  We now have self.scheduler.restart() and self.server.schedulers['name'].restart().  You should have added a dictionary of schedulers as part of self.

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc1")
+        a = {'comment': ''}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
```
I think op=UNSET will try to check if the comment attribute is removed . here comment is set to empty string

bhroam replied:
Why would the comment be an empty string?  It should either be set to something or not show up in the qmgr output.

visheshh replied:
@suresh-thelkar: can you answer the question ?

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+            self.du.run_copy(self.server.hostname,
+                             os.path.join(pbs_home, 'sched_logs'),
+                             os.path.join(sched_home, sched_logs_dir),
+                             recursive=True)
```
I have to create the directory . Okay i will change it to mkdir .

visheshh commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -0,0 +1,601 @@
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id="sc1")
+        a = {'comment': ''}
+        self.server.expect(SCHED, a, id='sc1', max_attempts=10)
```
I think op=UNSET will try to check if the comment attribute is removed . here comment is set to empty string

arungrover commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,358 @@ next_job(status *policy, server_info *sinfo, int flag)
+ *
+ */
+static int
+copy_attr_value(char **dest, char *src)
```
you can either make it more generic and call it pbs_strcpy and put it in misc_utils.c or just make use of pbs_strcat to the same thing (just reset the destination to '\0' everytime before copy.

arungrover commented on src/scheduler/fifo.c:
```diff
@@ -2327,3 +2350,358 @@ next_job(status *policy, server_info *sinfo, int flag)
+			} else if (!strcmp(attr->name, ATTR_partition)) {
+				if (copy_attr_value(&tmp_partitions, attr->value) == 0)
+					return 0;
+			} else if (!strcmp(attr->name, ATTR_comment)) {
```
I'm not sure if I have asked this before, but why do we need a comment to be stored in scheduler?

bhroam commented:
This pull request is getting unresponsive.  I'm going to close it.  Please open a new pull request and reference this one.  This is an unfortunate side effect of a pull request growing too large.

suresh-thelkar commented on src/lib/Liblog/pbs_messages.c:
```diff
@@ -405,6 +405,8 @@ char *msg_cannot_set_route_que = "Route queues are incompatible with the partiti
 char *msg_partition_not_in_queue = "Partition %s is not part of queue for node";
 char *msg_invalid_partion_in_queue = "Invalid partition in queue";
+char *msg_sched_op_not_permitted = "Operation is not permitted on default scheduler";
+char *msg_sched_part_already_used = "Partition is already associated with other scheduler";
```
As of today let us say if somebody wants to query a list of partitions that are present we do not have a direct way of 
 knowing them since partition is logical attribute which relates queues/nodes/schedulers. We actually need to query all the queues/nodes/schedulers to get the list of partitions. I think it is better to add this ability first if there is such a requirement for this and then come up with these elaborate messages.  As there are already lot of changes in the same ticket, planning to take both of these as part of another ticket after discussing the these requirement/s in detail.
